File: go/day_3/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func main() {
	partOne()
	partTwo()
}

const tree = byte('#')

func partOne() {
	fmt.Println("Part One:", countTrees(3, 1))
	return
}

func partTwo() {
	treeCount := 1

	moves := [5][2]int{
		{1, 1},
		{3, 1},
		{5, 1},
		{7, 1},
		{1, 2},
	}

	for i := 0; i < len(moves); i++ {
		treeCount *= countTrees(moves[i][0], moves[i][1])
	}

	fmt.Println("Part Two:", treeCount)
	return
}

func countTrees(rightMoves, downMoves int) int {
	file, err := os.Open("trees.txt")
	check(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	treeCount := 0

	i := 0
	row := 0
	for scanner.Scan() {
		row++
		if row%downMoves == 1 {
			continue
		}

		if scanner.Text()[i] == tree {
			treeCount++
		}

		i = mod(i+rightMoves, 31)
	}

	return treeCount
}

func mod(a, b int) int {
	m := a % b
	if a < 0 && b < 0 {
		m -= b
	}
	if a < 0 && b > 0 {
		m += b
	}
	return m
}

func check(err error) {
	if err != nil {
		log.Fatal(err)
	}
}

File: rust/day_3/src/main.rs
use std::fs::File;
use std::io::{prelude::*, BufReader};

const TREE: char = '#';

fn main() {
    part_one();
    part_two();
}

fn part_one() {
    println!("Part One: {}", count_trees(3, 1))
}

fn part_two() {
    const MOVES: [[usize; 2]; 5] = [
        [1, 1],
        [3, 1],
        [5, 1],
        [7, 1],
        [1, 2],
    ];
    let mut tree_count: u32 = 1;

    for i in 0..5 {
        tree_count = tree_count * count_trees(MOVES[i][0], MOVES[i][1]);
    }

    println!("Part Two: {}", tree_count);
}

fn count_trees(right_moves: usize, down_moves: usize) -> u32 {
    let file = File::open("trees.txt").expect("");
    let reader = BufReader::new(file);
    
    let mut tree_count: u32 = 0;
    let mut col: usize = 0;
    let mut row: usize = 0;  

    for line in reader.lines() {
        row = row + 1;
        if row%down_moves == 1 {
            continue
        }

        if line.unwrap().chars().nth(col).unwrap() == TREE {
            tree_count = tree_count + 1
        }
        
        col = modulo(col+right_moves, 31)
    }

    return tree_count;
}

fn modulo(a: usize, b: usize) -> usize {
    return ((a % b) + b) % b
}